### PR TITLE
API Application.Deploy: accept an args object.

### PR DESCRIPTION
**Pull request by frankban**
Also handle pending charm resources.

### CODE REVIEW

**Comment by frankban** [comment]
QA: as this deeply changes how applications are deployed, please make sure deploying charms works in both sandbox mode and gijoe. All ecs actions should work, including adding pending relations, more units, changing config/exposing before committing etc.

**Comment by jujugui** [comment]
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
http://ci.jujugui.org:8080//job/juju-gui/6146/
Test PASSed.

File: jujugui/static/gui/src/app/utils/bundle-importer.js [hunk 1]
```diff
@@ -482,36 +482,36 @@ YUI.add('bundle-importer', function(Y) {
...
+            this.modelAPI.update_annotations(
+              name, 'application', ghostService.get('annotations'));
+          }.bind(this, ghostService);
+          this.modelAPI.deploy({
+            charmURL: charm.get('id'),
+            applicationName: record.args[2],
+            series: series,
+            config: record.args[3],
+            constraints: constraints
+          }, deployCallback, {modelId: ghostService.get('id')});
```
Comment (hatched) [inline jujugui/static/gui/src/app/utils/bundle-importer.js]: Nice, this is cleaner. Will we ever have to provide resources here? Just curious.

File: jujugui/static/gui/src/app/utils/environment-change-set.js [hunk 2]
```diff
@@ -617,50 +617,43 @@ YUI.add('environment-change-set', function(Y) {
...
           // Get the key to the record which adds the charm for this app.
-          if (record.command.args[0] === args[0]) {
+          if (record.command.args[0] === args[0].charmURL) {
             parents.push(key);
           }
         }
       });
-      // The 6th param is the toMachine param of the env deploy call.
-      var toMachine = command.args[7];
-      if (!this.changeSet[toMachine]) {
```
Comment (hatched) [inline jujugui/static/gui/src/app/utils/environment-change-set.js]: Why isn't this necessary any longer?

File: jujugui/static/gui/src/app/utils/bundle-importer.js (hunk 1 above)
Comment (frankban) [inline jujugui/static/gui/src/app/utils/bundle-importer.js]: Not sure.

File: jujugui/static/gui/src/app/utils/environment-change-set.js (hunk 2 above)
Comment (frankban) [inline jujugui/static/gui/src/app/utils/environment-change-set.js]: It's never used, we only place units when calling addUnit, not on deploy calls.

**Comment by hatched** [comment]
Thanks @frankban :shipit:

**Comment by jujugui** [comment]
Status: merge request accepted. Url: http://ci.jujugui.org:8080/job/juju-gui-merge

**Pull request merged**